### PR TITLE
[SPARK-17491] Close serialization stream to fix wrong answer bug in putIteratorAsBytes()

### PR DESCRIPTION
**Pull request by JoshRosen**
## What changes were proposed in this pull request?

`MemoryStore.putIteratorAsBytes()` may silently lose values when used with `KryoSerializer` because it does not properly close the serialization stream before attempting to deserialize the already-serialized values, which may cause values buffered in Kryo's internal buffers to not be read.

This is the root cause behind a user-reported "wrong answer" bug in PySpark caching reported by @bennoleslie on the Spark user mailing list in a thread titled "pyspark persist MEMORY_ONLY vs MEMORY_AND_DISK". Due to Spark 2.0's automatic use of KryoSerializer for "safe" types (such as byte arrays, primitives, etc.) this misuse of serializers manifested itself as silent data corruption rather than a StreamCorrupted error (which you might get from JavaSerializer). 

The minimal fix, implemented here, is to close the serialization stream before attempting to deserialize written values. In addition, this patch adds several additional assertions / precondition checks to prevent misuse of `PartiallySerializedBlock` and `ChunkedByteBufferOutputStream`.
## How was this patch tested?

The original bug was masked by an invalid assert in the memory store test cases: the old assert compared two results record-by-record with `zip` but didn't first check that the lengths of the two collections were equal, causing missing records to go unnoticed. The updated test case reproduced this bug.

In addition, I added a new `PartiallySerializedBlockSuite` to unit test that component.

### CODE REVIEW

**Comment by lins05** [comment]
Did a simple test and it does fix the bug. One interesting thing is while records.count() returns a smaller number than the actual count, the spark UI still shows the correct records number, in my test case it's 2999808 v.s. 300000.

screen shot 2016-09-10 at 5 57 38 pm

screen shot 2016-09-10 at 5 58 26 pm

**Comment by JoshRosen** [comment]
#15056 also touches this code and creates a new test suite for this component so I'd prefer to merge that PR first.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 1]
```diff
@@ -782,6 +785,9 @@ private[storage] class PartiallySerializedBlock[T](
    * `close()` on it to free its resources.
    */
   def valuesIterator: PartiallyUnrolledIterator[T] = {
+    // Close the serialization stream so that the serializer's internal buffers are freed and any
+    // "end-of-stream" markers can be written out so that `unrolled` is a valid serialized stream.
+    serializationStream.close()
```
Comment (srinathshankar) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: It seems like 'unrolled' may basically be invalid until serializationStream is called.

But it looks like valuesIterator is not the only place where unrolled is used

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 2]
```diff
@@ -794,13 +821,17 @@ private[storage] class PartiallySerializedBlock[T](
...
     // `unrolled`'s underlying buffers will be freed once this input stream is fully read:
     val unrolledIter = serializerManager.dataDeserializeStream(
-      blockId, unrolled.toInputStream(dispose = true))(classTag)
+      blockId, unrolledBuffer.toInputStream(dispose = true))(classTag)
     new PartiallyUnrolledIterator(
       memoryStore,
       unrollMemory,
-      unrolled = CompletionIterator[T, Iterator[T]](unrolledIter, discard()),
+      unrolled = CompletionIterator[T, Iterator[T]](unrolledIter, unrolledBuffer.dispose()),
```
Comment (srinathshankar) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: Why the change from discard to dispose() ? You've made discard idempotent, right ? Does the caller have to manually release memory after the iterator is consumed ?

File: core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    test(s"$testCaseName with valuesIterator() and numBuffered = $numItemsToBuffer") {
+      val partiallySerializedBlock = partiallyUnroll(items.iterator, numItemsToBuffer)
+      val valuesIterator = partiallySerializedBlock.valuesIterator
+      Mockito.verify(partiallySerializedBlock.invokePrivate(getSerializationStream())).close()
+      Mockito.verify(partiallySerializedBlock.invokePrivate(getRedirectableOutputStream())).close()
+
+      val deserializedItems = valuesIterator.toArray.toSeq
+      Mockito.verify(memoryStore).releaseUnrollMemoryForThisTask(
```
Comment (srinathshankar) [inline core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala]: What's the code path that makes releaseUnroll be called, give that the completion task returned by valuesIterator has changed from discard to dispose .

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala (hunk 2 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: There was a subtle bug in the old code where the call to `dispose()` would end up freeing unroll memory for the buffer but that same memory would also be freed by `PartiallyUnrolledIterator` itself. This was exposed by the Mockito verify calls in the new test suite.

Given that `unrolledBuffer.toInputStream(dispose = true)` will also handle disposal of the buffer I don't think we even need this `CompletionIterator` here. Let me see about removing it.

File: core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    val unrollMemory = bbos.size
+    new PartiallySerializedBlock[T](
+      memoryStore,
+      serializerManager,
+      blockId,
+      serializationStream = serializationStream,
+      redirectableOutputStream,
+      unrollMemory = unrollMemory,
+      memoryMode = MemoryMode.ON_HEAP,
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala]: It should only affect the memory accounting for unroll memory. If you're caching a block at `MEMORY_SER` storage level and are using off-heap caching then it's possible for the unrolled memory to be off-heap (so the ChunkedByteBufferOutputStream will be using a DirectBuffer allocator). In this case we need to count this as off-heap unroll memory so that Spark's off-heap allocations can respect the configured off-heap memory limit.

Given that off-heap caching (and thus off-heap unrolling) is a relatively new experimental feature, it's entirely possible that there are accounting bugs within this path. I'm going to try to expand this test suite to also exercise that case just to be 100% sure that we're accounting properly.

File: core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala (hunk 3 above)
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala]: Any non-freed unroll memory will be automatically freed at the end of the task (as part of the `Executor` or `Task` code itself). Before the task has completed, though, `PartiallyUnrolledIterator` will free the specified amount of unroll memory once the unrolled iterator is fully consumed.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 5]
```diff
@@ -751,23 +757,43 @@ private[storage] class PartiallySerializedBlock[T](
...
+
+  private def verifyNotConsumedAndNotDiscarded(): Unit = {
+    if (consumed) {
+      throw new IllegalStateException(
+        "Can only call one of finishWritingToStream() or valuesIterator() and can only call once.")
+    }
+    if (discarded) {
+      throw new IllegalStateException("Cannot call methods on a discarded PartiallySerializedBlock")
     }
+    consumed = true
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala]: I was being overly-clever; it's clearer to just set `consumed = true` after each of the `verifyNotConsumedAndNotDiscarded` call sites.

**Review by JoshRosen — COMMENTED** [review]
Thanks for the good review feedback. I think that there might indeed be a pre-existing bug related to off-heap unroll accounting here, so let me try to also catch that via strengthened test cases. I'll update this patch to address your feedback.

**Comment by JoshRosen** [comment]
Alright, I've updated this to address the latest round of review feedback. I did manage to spot a memory-accounting problem with off-heap memory because `PartiallyUnrolledIterator` had hardcoded the use of `MemoryMode.OFF_HEAP`.

If this change

File: core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  test("cannot call finishWritingToStream() after valuesIterator()") {
+    val partiallySerializedBlock = partiallyUnroll((1 to 10).iterator, 2)
+    partiallySerializedBlock.valuesIterator
+    intercept[IllegalStateException] {
+      partiallySerializedBlock.finishWritingToStream(new ByteBufferOutputStream())
+    }
+  }
+
+  test("cannot call valuesIterator() after finishWritingToStream()") {
```
Comment (srinathshankar) [inline core/src/test/scala/org/apache/spark/storage/PartiallySerializedBlockSuite.scala]: Minor: You can probably combine 
test("cannot call valuesIterator() more than once") and
test("cannot call finishWritingToStream() after valuesIterator()") into one test
Same for making calls after finishWritingToStream

File: core/src/test/scala/org/apache/spark/storage/PartiallyUnrolledIteratorSuite.scala [hunk 7]
```diff
@@ -33,7 +33,7 @@ class PartiallyUnrolledIteratorSuite extends SparkFunSuite with MockitoSugar {
     val rest = (unrollSize until restSize + unrollSize).iterator
 
     val memoryStore = mock[MemoryStore]
-    val joinIterator = new PartiallyUnrolledIterator(memoryStore, unrollSize, unroll, rest)
+    val joinIterator = new PartiallyUnrolledIterator(memoryStore, ON_HEAP, unrollSize, unroll, rest)
```
Comment (srinathshankar) [inline core/src/test/scala/org/apache/spark/storage/PartiallyUnrolledIteratorSuite.scala]: We should look into trying to test this with OFF_HEAP as well.

**Review by srinathshankar — COMMENTED** [review]
This looks good. I think a test dimension with memory_mode = off_heap would be useful

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
I agree that we should add more off-heap tests, but I'd like to do it in another patch so that we can get this one merged faster to unblock the 2.0.1 RC.

In terms of testing off-heap, I think that one of the best high-level tests / asserts would be to strengthen the `releaseUnrollMemory()` checks so that inappropriately releasing unroll memory _during_ a task throws an exception during tests. Today there are some circumstances where unroll memory can only be released at the end of a task (such as an iterator backed by an unrolled block that is only partially consumed before the task ends), so the calls to release unroll memory have been tolerant of too much memory being released (it just releases `min(actualMemory, requestedToRelease)`). However, this is only appropriate to do at the end of the task so we should strengthen the asserts to only allow it there; this would have caught the memory mode mixup that I fixed here.

I'm going to retest this and if it passes tests then I'll merge to master and branch-2.0. I'll add the new tests described above in a followup.

**Comment by JoshRosen** [comment]
I believe that this latest test failure is caused by a known flaky PySpark test, so I'm going to merge this now and will monitor tests afterwards.
